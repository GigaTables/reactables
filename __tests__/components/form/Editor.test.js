import React from 'react'
import renderer from 'react-test-renderer'
import Editor from '../../../src/components/form/Editor.js'
import { shallow, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

configure({ adapter: new Adapter() })

const CommonConstants = require('../../../src/components/CommonConstants')
const EditorConstants = require('../../../src/components/EditorConstants')

const editor = {
    ajax: 'http://gigatables.loc/editor.php',
    ajaxFiles: 'uploadFiles.php',
    struct: {
        buttons: [ 'top', 'bottom' ] // buttons
    },
    fields: [
        {
            label: 'ID',
            name: 'id',
            type: 'hidden'
        },
        {// an example of using select - automatically selected if matches with data in table column
            label: 'Types:',
            name: 'types[]',
            values: [// if select,checkbox,radio etc types - need this pre-set structure of values
                { 'key1': 'val1' },
                { 'key2': 'val2' }
            ],
            type: 'checkbox' // select,checkbox,radio
        },
        {// an example of using select - automatically selected if matches with data in table column
            label: 'Cities:',
            name: 'cities[]',
            values: [// if select,checkbox,radio etc types - need this pre-set structure of values
                { 'key1': 'val1' },
                { 'key2': 'val2' }
            ],
            type: 'radio' // select,checkbox,radio
        },
        {// an example of using select - automatically selected if matches with data in table column
            label: 'Types:',
            name: 'types[]',
            values: [// if select,checkbox,radio etc types - need this pre-set structure of values
                { 'key1': 'val1' },
                { 'key2': 'val2' }
            ],
            type: 'select' // select,checkbox,radio
//              attrs: [
//                {'multiple':true}
//              ]
        },
        {
            label: 'Article title:',
            name: 'title',
            type: 'text', // default, other: password, file, select, multiselect etc
            attrs: [
                { 'pattern': '^[A-Za-z0-9_]+$' }
            ]
        },
        {
            label: 'Description:',
            name: 'desc',
            type: 'textarea'
        },
        {
            label: 'Description:',
            name: 'desc',
            type: 'textarea',
            plugins: 'rte',
            attrs: [
                { 'className': 'descriptionField' }
            ],
            fieldsetClose: true
        },
        {
            label: 'Date Time:',
            name: 'date',
            type: 'date'
        },
        {
            label: 'Image:',
            name: 'image',
            type: 'file'
        }
    ]
}

const settings = {
    struct: {
        search: [ 'top', 'bottom' ],
        rowsSelector: [ 'desc', 'top', 'bottom' ],
        pagination: [ 'bottom' ]
    },
    lang: 'en', // english default
    perPageRows: [ 25, 50, 100, 200 ],
    defaultPerPage: 100,
    ajax: 'http://gigatables.loc/gigatables.php',
    // ajaxAutoloadData: true, // default false
    // ajaxAutoloadPeriod: 8, // sec
    requestType: 'GET',
    columns: [
        {// include all defaults
            data: 'id',
            sortable: true, // true by defualt
            visible: true, // true by defualt
            searchable: true, // true by defualt
            discreteSearch: true, // false by default
            discreteSearchValue: function (title) {
                return 'Search by field - ' + title
            }
        },
        {
            data: 'title',
            cISearch: true // default false
        },
        {
            data: 'desc',
            sortable: false,
            discreteSearch: true,
            discreteCISearch: true // default false
        },
        {
            data: 'date',
            searchable: false
        },
        {
            data: 'info'
        },
        { data: 'field1' },
        { data: 'field2' },
        { data: 'field3', visible: false }
    ],
    columnOpts: [
        {
            render: (data, row, type) => (
                <div>
                    <form method="post" action="">
                        <input type="hidden" name="action" value="forms"/>
                        <input type="hidden" name="id" value={row.id}/>
                        <div>{data}</div>
                    </form>
                </div>),
            target: 'title' // provide data column index to match opts
        },
        {
            render: (data, row, type) => (
                <div>
                    <div>{data}</div>
                    <div>{(row.id % 2 === 0) ? '**' : '*'}</div>
                </div>
            ),
            target: 'date'
        }
    ],
    tableOpts: {
        buttons: [
            {
                extended: 'editor_create', editor: editor, triggerAfter: (function () {
                console.log('after create')
            }), triggerBefore: (function () {
                console.log('before create')
            })
            },
            {
                extended: 'editor_edit', editor: editor, triggerBefore: (function () {
                console.log('before edit')
            })
            },
            {
                extended: 'editor_remove', editor: editor, triggerAfter: (function () {
                console.log('after del')
            })
            }
        ],
        buttonsPosition: [ 'top', 'bottom' ],
        theme: 'material-ui'
    }
}

describe('Editor', () => {
    
    beforeEach(() => {
        global.editorUpdate = jest.fn().mockImplementation((e, dataResp) => {
        
        })
        global.fetch = jest.fn().mockImplementation(() => {
            var p = new Promise((resolve, reject) => {
                resolve({
                    ok: true,
                    Id: '123',
                    json: function () {
                        return {
                            'rows': [
                                {
                                    'GT_RowId': 123,
                                    'title': 'Test Bar 123st row',
                                    'id': 123,
                                    'desc': 'Lorem Ipsum is simply dummy Bar 7196 text of the printing and typesetting',
                                    'info': 'some info some info some info some info',
                                    'date': '13:23:43 02:04:2017',
                                    'field1': 123,
                                    'field2': 1357,
                                    'field3': 12468
                                }
                            ]
                        }
                    }
                })
            })
            return p
        })
    })
    
    it('renders Editor correctly', () => {
        const tree = renderer.create(
            <Editor
                lang="en"
                action="create"
                countRows="229"
                fromRow="0"
                page="1"
                perPage="50"
                active={true}
                editor={editor}
                columns={editor.fields}
                struct={settings.struct}
                tableOpts={settings.tableOpts}
                selectedRows={[ 1 ]}
                selectedIds={[ 187 ]}
                fieldsEdit={[
                    { 'id': 1 }
                ]}
                editorUpdate={(e, resp) => {}}/>
        ).toJSON()
        expect(tree).toMatchSnapshot()
        
        const obj = shallow(
            <Editor
                lang="en"
                countRows="229"
                fromRow="0"
                page="1"
                perPage="50"
                action="delete"
                active={true}
                editor={editor}
                columns={editor.fields}
                struct={settings.struct}
                tableOpts={settings.tableOpts}
                selectedRows={[ 1 ]}
                selectedIds={[ 187 ]}
                fieldsEdit={[
                    { 'id': 1 }
                ]}
                editorUpdate={(e, resp) => {}}/>
        )
        obj.instance().setDataIndices({
            columns: settings.columns
        })
        obj.instance().setFields({
            action: 'create',
            editor: editor
        })
        obj.instance().setFields({
            action: 'edit',
            editor: editor
        })
        obj.instance().setFields({
            action: 'delete',
            editor: editor,
            selectedRows: [ 1, 2, 3 ]
        })
        obj.instance().triggerBefore({
            type: 'editor_create'
        })
        obj.instance().triggerAfter({
            type: 'editor_remove'
        })
        obj.instance().onChange({
            target: {
                name: 'foo',
                value: 'bar',
                dataset: {
                    multiple: false,
                    textarea: 'true'
                }
            }
        })
        obj.instance().onFocus({
            target: {
                dataset: {
                    multiple: false,
                    textarea: 'true'
                },
                children: [] // todo: set DOM for attr
            }
        })
        obj.instance().stopPropagation({
            stopPropagation: () => {
            }
        })
        obj.instance().btnClickedEnter({
            keyCode: CommonConstants.ENTER_KEY
        })
        obj.instance().getAjaxSettings(EditorConstants.ACTION_EDIT)
        obj.instance().getAjaxSettings(EditorConstants.ACTION_DELETE)
        // todo: resolve error: Headers is not defined
        // global.Headers = () => {}
        obj.instance().btnClicked({
            persist: () => {
            }
        })
    })
    it('renders Editor with ajax object', () => {
        editor.ajax = {
            create: {
                url: 'http://gigatables.loc/editor.php',
                type: 'POST',
                headers: {
                    'X-Api-Key': '8013b37216a07f50027139d89ee9f822e3784049'
                }
            },
            edit: {
                url: 'http://gigatables.loc/editor.php',
                type: 'PUT',
                headers: {
                    'X-Api-Key': '8013b37216a07f50027139d89ee9f822e3784049',
                    'X-Header-Key': 'foo-bar'
                }
            },
            delete: {
                url: 'http://gigatables.loc/editor.php',
                type: 'DELETE',
                headers: {
                    'X-Api-Key': '8013b37216a07f50027139d89ee9f822e3784049',
                    'X-Header-Key': 'foo-bar-baz'
                }
            }
        }
        const ajaxObject = shallow(
            <Editor
                lang="en"
                countRows="229"
                fromRow="0"
                page="1"
                perPage="50"
                action="edit"
                active={true}
                editor={editor}
                columns={editor.fields}
                struct={settings.struct}
                tableOpts={settings.tableOpts}
                selectedRows={[ 1 ]}
                selectedIds={[ 187 ]}
                fieldsEdit={[
                    { 'id': 1 }
                ]}
                editorUpdate={(e, resp) => {}}/>
        )
        ajaxObject.instance().getAjaxSettings('edit')
        expect(() => {ajaxObject.instance().getAjaxSettings('dummy')}).toThrow()
        ajaxObject.instance().setHeaders(
            {
                headers: {
                    'X-Api-Key': '8013b37216a07f50027139d89ee9f822e3784049',
                    'X-Header-Key': 'foo-bar'
                }
            }, { append: (key, val) => {} })
        ajaxObject.instance().btnClicked({
            persist: () => {
            }
        })
    })
})

describe('Editor create with AJAX row object returned', () => {
    beforeEach(() => {
        global.editorUpdate = jest.fn().mockImplementation((e, dataResp) => {
        
        })
        global.fetch = jest.fn().mockImplementation(() => {
            var p = new Promise((resolve, reject) => {
                resolve({
                    ok: true,
                    Id: '123',
                    json: function () {
                        return {
                            'row':
                                {
                                    'GT_RowId': 123,
                                    'title': 'Test Bar 123st row',
                                    'id': 123,
                                    'desc': 'Lorem Ipsum is simply dummy Bar 7196 text of the printing and typesetting',
                                    'info': 'some info some info some info some info',
                                    'date': '13:23:43 02:04:2017',
                                    'field1': 123,
                                    'field2': 1357,
                                    'field3': 12468
                                }
                        }
                    }
                })
            })
            return p
        })
    })
    
    it('renders Editor with ajax object', () => {
        const ajaxObjectCreate = shallow(
            <Editor
                lang="en"
                countRows="229"
                fromRow="0"
                page="1"
                perPage="50"
                action="create"
                active={true}
                editor={editor}
                columns={editor.fields}
                struct={settings.struct}
                tableOpts={settings.tableOpts}
                selectedRows={[ 1 ]}
                selectedIds={[ 187 ]}
                fieldsEdit={[
                    { 'id': 1 }
                ]}
                editorUpdate={(e, resp) => {}}/>
        )
        ajaxObjectCreate.instance().btnClicked({
            persist: () => {
            }
        })
    })
})

describe('Editor create with AJAX row object returned', () => {
    beforeEach(() => {
        global.editorUpdate = jest.fn().mockImplementation((e, dataResp) => {
        
        })
        global.fetch = jest.fn().mockImplementation(() => {
            var p = new Promise((resolve, reject) => {
                resolve({
                    ok: true,
                    Id: '123',
                    json: function () {
                        return {
                            'row':
                                {
                                    'GT_RowId': 123,
                                    'title': 'Test Bar 123st row',
                                    'id': 123,
                                    'desc': 'Lorem Ipsum is simply dummy Bar 7196 text of the printing and typesetting',
                                    'info': 'some info some info some info some info',
                                    'date': '13:23:43 02:04:2017',
                                    'field1': 123,
                                    'field2': 1357,
                                    'field3': 12468
                                }
                        }
                    }
                })
            })
            return p
        })
    })
    
    it('renders Editor with ajax object', () => {
        const ajaxObjectCreate = shallow(
            <Editor
                lang="en"
                countRows="229"
                fromRow="0"
                page="1"
                perPage="50"
                action="create"
                active={true}
                editor={editor}
                columns={editor.fields}
                struct={settings.struct}
                tableOpts={settings.tableOpts}
                selectedRows={[ 1 ]}
                selectedIds={[ 187 ]}
                fieldsEdit={[
                    { 'id': 1 }
                ]}
                editorUpdate={(e, resp) => {}}/>
        )
        ajaxObjectCreate.instance().setState({
            dataIndices: {
                'GT_RowId': 123,
                'id': 123,
                'title': 'foo'
            }
        })
        ajaxObjectCreate.instance().onChange({
            target: {
                name: 'title',
                value: 'bar',
                dataset: {
                    multiple: false,
                    textarea: 'true'
                }
            }
        })
        ajaxObjectCreate.instance().setDataIndices({
            columns: settings.columns
        })
        ajaxObjectCreate.instance().btnClicked({
            persist: () => {
            }
        })
    })
})

describe('Editor throws exceptions on', () => {
    beforeEach(() => {
        global.editorUpdate = jest.fn().mockImplementation((e, dataResp) => {
        
        })
        global.fetch = jest.fn().mockImplementation(() => {
            let p = new Promise((resolve, reject) => {
                resolve({
                    ok: true,
                    Id: '123',
                    json: function () {
                        return {
                            'row':
                                {
                                    // 'GT_RowId': 123,
                                    'title': 'Test Bar 123st row',
                                    // 'id': 123,
                                    'desc': 'Lorem Ipsum is simply dummy Bar 7196 text of the printing and typesetting',
                                    'info': 'some info some info some info some info',
                                    'date': '13:23:43 02:04:2017',
                                    'field1': 123,
                                    'field2': 1357,
                                    'field3': 12468
                                }
                        }
                    }
                })
            })
            return p
        })
    })
    
    it('create without row', () => {
        const ajaxObjectCreate = shallow(
            <Editor
                lang="en"
                countRows="229"
                fromRow="0"
                page="1"
                perPage="50"
                action="create"
                active={true}
                editor={editor}
                columns={editor.fields}
                struct={settings.struct}
                tableOpts={settings.tableOpts}
                selectedRows={[ 1 ]}
                selectedIds={[ 187 ]}
                fieldsEdit={[
                    { 'id': 1 }
                ]}
                editorUpdate={(e, resp) => {}}/>
        )
        expect(ajaxObjectCreate.instance().btnClicked({
            persist: () => {
            }
        })).toThrow()
    })
    
    it('edit  without rows', () => {
        const ajaxObjectCreate = shallow(
            <Editor
                lang="en"
                countRows="229"
                fromRow="0"
                page="1"
                perPage="50"
                action="edit"
                active={true}
                editor={editor}
                columns={editor.fields}
                struct={settings.struct}
                tableOpts={settings.tableOpts}
                selectedRows={[ 1 ]}
                selectedIds={[ 187 ]}
                fieldsEdit={[
                    { 'id': 1 },
                    { 'id': 2 }
                ]}
                editorUpdate={(e, resp) => {}}/>
        )
        expect(ajaxObjectCreate.instance().btnClicked({
            persist: () => {
            }
        })).toThrow()
        ajaxObjectCreate.setState({
            setMultipleText: 1
        })
        ajaxObjectCreate.instance().getFieldByType('desc', { name: 'title', type: 'text', label: '' })
        ajaxObjectCreate.setState({
            dataIndices: {
                'GT_RowId': 123,
                'id': 123,
                'title': 'foo'
            }
        })
        ajaxObjectCreate.instance().getFieldByType('desc', { name: 'title', type: 'text', label: '' })
        ajaxObjectCreate.instance().onChangeHtml({
            props: {
                name: 'desc'
            }
        }, '<div>foo bar baz</div>')
    })
})
