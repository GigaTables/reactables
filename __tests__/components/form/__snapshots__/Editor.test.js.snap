exports[`Editor renders Editor correctly 1`] = `
<div>
  <div
    className="gte_editor_popup fade_in"
    onClick={undefined}>
    <div
      className="gte_popup_container">
      <div
        className="gte_popup_container_wrapper">
        <div
          className="gte_form_border_box"
          onClick={[Function]}
          onKeyUp={[Function]}>
          <div
            className="close_btn"
            onClick={[Function]} />
          <div
            className="gte_form_fields">
            <div
              className="gte_header">
              <div
                className="gte_editor_title" />
            </div>
            <div
              className="gte_form_body">
              <div
                className="gte_form_body_content">
                <form
                  action=""
                  id="gte_form"
                  method="post">
                  <div
                    className="gte_form_content">
                    <div>
                      <input
                        name="action"
                        type="hidden"
                        value="create" />
                    </div>
                    <div>
                      <div
                        className="gte_editor_fields">
                        <fieldset>
                          <legend>
                            
                          </legend>
                          <div
                            className="gte_editor_fields">
                            <label
                              className="gte_label"
                              htmlFor="id" />
                            <div
                              className={undefined}>
                              <input
                                data-multiple={false}
                                id="id"
                                name="id"
                                onChange={[Function]}
                                onFocus={[Function]}
                                type="hidden"
                                value="" />
                            </div>
                            <div
                              className="clear" />
                          </div>
                          <div
                            className="gte_editor_fields">
                            <label
                              className="gte_label"
                              htmlFor="types[]">
                              Types:
                            </label>
                            <div
                              className={undefined}>
                              <label
                                className="gte_label_text">
                                <input
                                  data-value="val1"
                                  defaultChecked={0}
                                  id="types"
                                  name="types[]"
                                  onClick={[Function]}
                                  type="checkbox"
                                  value="key1" />
                                val1
                              </label>
                              <label
                                className="gte_label_text">
                                <input
                                  data-value="val2"
                                  defaultChecked={0}
                                  id="types"
                                  name="types[]"
                                  onClick={[Function]}
                                  type="checkbox"
                                  value="key2" />
                                val2
                              </label>
                            </div>
                            <div
                              className="clear" />
                          </div>
                          <div
                            className="gte_editor_fields">
                            <label
                              className="gte_label"
                              htmlFor="types[]">
                              Types:
                            </label>
                            <div
                              className={undefined}>
                              <select
                                id="types[]"
                                name="types[]"
                                onChange={[Function]}
                                value="">
                                <option
                                  data-value="val1"
                                  value="key1">
                                  val1
                                </option>
                                <option
                                  data-value="val2"
                                  value="key2">
                                  val2
                                </option>
                              </select>
                            </div>
                            <div
                              className="clear" />
                          </div>
                          <div
                            className="gte_editor_fields">
                            <label
                              className="gte_label"
                              htmlFor="title">
                              Article title:
                            </label>
                            <div
                              className={undefined}>
                              <input
                                0={
                                  Object {
                                    "pattern": "^[A-Za-z0-9_]+$",
                                  }
                                }
                                data-multiple={false}
                                id="title"
                                name="title"
                                onChange={[Function]}
                                onFocus={[Function]}
                                type="text"
                                value="" />
                            </div>
                            <div
                              className="clear" />
                          </div>
                          <div
                            className="gte_editor_fields">
                            <label
                              className="gte_label"
                              htmlFor="desc">
                              Description:
                            </label>
                            <div
                              className={undefined}>
                              <textarea
                                data-multiple={undefined}
                                data-textarea={true}
                                id="desc"
                                name="desc"
                                onChange={[Function]}
                                onFocus={[Function]}
                                value="" />
                            </div>
                            <div
                              className="clear" />
                          </div>
                          <div
                            className="gte_editor_fields">
                            <label
                              className="gte_label"
                              htmlFor="desc">
                              Description:
                            </label>
                            <div
                              className={undefined}>
                              <div
                                className="RichTextEditor__root___2QXK-"
                                style={undefined}>
                                <div
                                  className="EditorToolbar__root___3_Aqz"
                                  style={undefined}>
                                  <div
                                    className="ButtonGroup__root___3lEAn">
                                    <div
                                      className="ButtonWrap__root___1EO_R">
                                      <button
                                        className="IconButton__root___3tqZW Button__root___1gz0c"
                                        disabled={undefined}
                                        onClick={[Function]}
                                        onMouseDown={[Function]}
                                        title="Bold"
                                        type="button">
                                        <span
                                          className="IconButton__icon-bold___2zl9t IconButton__icon___3YgOS" />
                                      </button>
                                    </div>
                                    <div
                                      className="ButtonWrap__root___1EO_R">
                                      <button
                                        className="IconButton__root___3tqZW Button__root___1gz0c"
                                        disabled={undefined}
                                        onClick={[Function]}
                                        onMouseDown={[Function]}
                                        title="Italic"
                                        type="button">
                                        <span
                                          className="IconButton__icon-italic___2hHzc IconButton__icon___3YgOS" />
                                      </button>
                                    </div>
                                    <div
                                      className="ButtonWrap__root___1EO_R">
                                      <button
                                        className="IconButton__root___3tqZW Button__root___1gz0c"
                                        disabled={undefined}
                                        onClick={[Function]}
                                        onMouseDown={[Function]}
                                        title="Strikethrough"
                                        type="button">
                                        <span
                                          className="IconButton__icon-strikethrough___QtE2X IconButton__icon___3YgOS" />
                                      </button>
                                    </div>
                                    <div
                                      className="ButtonWrap__root___1EO_R">
                                      <button
                                        className="IconButton__root___3tqZW Button__root___1gz0c"
                                        disabled={undefined}
                                        onClick={[Function]}
                                        onMouseDown={[Function]}
                                        title="Monospace"
                                        type="button">
                                        <span
                                          className="IconButton__icon-code___3F1pe IconButton__icon___3YgOS" />
                                      </button>
                                    </div>
                                    <div
                                      className="ButtonWrap__root___1EO_R">
                                      <button
                                        className="IconButton__root___3tqZW Button__root___1gz0c"
                                        disabled={undefined}
                                        onClick={[Function]}
                                        onMouseDown={[Function]}
                                        title="Underline"
                                        type="button">
                                        <span
                                          className="IconButton__icon-underline___2EmZJ IconButton__icon___3YgOS" />
                                      </button>
                                    </div>
                                  </div>
                                  <div
                                    className="ButtonGroup__root___3lEAn">
                                    <div
                                      className="ButtonWrap__root___1EO_R">
                                      <button
                                        className="IconButton__root___3tqZW Button__root___1gz0c"
                                        disabled={undefined}
                                        onClick={[Function]}
                                        onMouseDown={[Function]}
                                        title="UL"
                                        type="button">
                                        <span
                                          className="IconButton__icon-unordered-list-item___Pvkrr IconButton__icon___3YgOS" />
                                      </button>
                                    </div>
                                    <div
                                      className="ButtonWrap__root___1EO_R">
                                      <button
                                        className="IconButton__root___3tqZW Button__root___1gz0c"
                                        disabled={undefined}
                                        onClick={[Function]}
                                        onMouseDown={[Function]}
                                        title="OL"
                                        type="button">
                                        <span
                                          className="IconButton__icon-ordered-list-item___2rzD0 IconButton__icon___3YgOS" />
                                      </button>
                                    </div>
                                    <div
                                      className="ButtonWrap__root___1EO_R">
                                      <button
                                        className="IconButton__root___3tqZW Button__root___1gz0c"
                                        disabled={undefined}
                                        onClick={[Function]}
                                        onMouseDown={[Function]}
                                        title="Blockquote"
                                        type="button">
                                        <span
                                          className="IconButton__icon-blockquote___17VSX IconButton__icon___3YgOS" />
                                      </button>
                                    </div>
                                  </div>
                                  <div
                                    className="ButtonGroup__root___3lEAn">
                                    <div
                                      className="ButtonWrap__root___1EO_R">
                                      <button
                                        className="IconButton__root___3tqZW Button__root___1gz0c"
                                        disabled={true}
                                        onClick={[Function]}
                                        onMouseDown={undefined}
                                        onSubmit={[Function]}
                                        title="Link"
                                        type="button">
                                        <span
                                          className="IconButton__icon-link___2umEl IconButton__icon___3YgOS" />
                                      </button>
                                    </div>
                                    <div
                                      className="ButtonWrap__root___1EO_R">
                                      <button
                                        className="IconButton__root___3tqZW Button__root___1gz0c"
                                        disabled={true}
                                        onClick={[Function]}
                                        onMouseDown={[Function]}
                                        title="Remove Link"
                                        type="button">
                                        <span
                                          className="IconButton__icon-remove-link___j61pw IconButton__icon___3YgOS" />
                                      </button>
                                    </div>
                                  </div>
                                  <div
                                    className="ButtonGroup__root___3lEAn">
                                    <div
                                      className="ButtonWrap__root___1EO_R">
                                      <button
                                        className="IconButton__root___3tqZW Button__root___1gz0c"
                                        disabled={undefined}
                                        onClick={[Function]}
                                        onMouseDown={undefined}
                                        onSubmit={[Function]}
                                        title="Image"
                                        type="button">
                                        <span
                                          className="IconButton__icon-image___1gW7U IconButton__icon___3YgOS" />
                                      </button>
                                    </div>
                                  </div>
                                  <div
                                    className="ButtonGroup__root___3lEAn">
                                    <span
                                      className="Dropdown__root___3ALmx"
                                      title="Normal">
                                      <select
                                        onChange={[Function]}
                                        value="unstyled">
                                        <option
                                          className={undefined}
                                          value="unstyled">
                                          Normal
                                        </option>
                                        <option
                                          className={undefined}
                                          value="header-one">
                                          Heading Large
                                        </option>
                                        <option
                                          className={undefined}
                                          value="header-two">
                                          Heading Medium
                                        </option>
                                        <option
                                          className={undefined}
                                          value="header-three">
                                          Heading Small
                                        </option>
                                        <option
                                          className={undefined}
                                          value="code-block">
                                          Code Block
                                        </option>
                                      </select>
                                      <span
                                        className="Dropdown__value___34Py9">
                                        Normal
                                      </span>
                                    </span>
                                  </div>
                                  <div
                                    className="ButtonGroup__root___3lEAn">
                                    <div
                                      className="ButtonWrap__root___1EO_R">
                                      <button
                                        className="IconButton__root___3tqZW Button__root___1gz0c"
                                        disabled={true}
                                        onClick={[Function]}
                                        onMouseDown={[Function]}
                                        title="Undo"
                                        type="button">
                                        <span
                                          className="IconButton__icon-undo___EQSRP IconButton__icon___3YgOS" />
                                      </button>
                                    </div>
                                    <div
                                      className="ButtonWrap__root___1EO_R">
                                      <button
                                        className="IconButton__root___3tqZW Button__root___1gz0c"
                                        disabled={true}
                                        onClick={[Function]}
                                        onMouseDown={[Function]}
                                        title="Redo"
                                        type="button">
                                        <span
                                          className="IconButton__icon-redo___30MVz IconButton__icon___3YgOS" />
                                      </button>
                                    </div>
                                  </div>
                                </div>
                                <div
                                  className="RichTextEditor__editor___1QqIU"
                                  style={undefined}>
                                  <div
                                    className="DraftEditor-root">
                                    <div
                                      className="DraftEditor-editorContainer">
                                      <div
                                        aria-activedescendant={undefined}
                                        aria-autocomplete={undefined}
                                        aria-describedby={null}
                                        aria-expanded={undefined}
                                        aria-haspopup={undefined}
                                        aria-label={undefined}
                                        aria-owns={undefined}
                                        autoCapitalize={undefined}
                                        autoComplete={undefined}
                                        autoCorrect={undefined}
                                        className="notranslate public-DraftEditor-content"
                                        contentEditable={true}
                                        data-testid={undefined}
                                        onBeforeInput={[Function]}
                                        onBlur={[Function]}
                                        onCompositionEnd={[Function]}
                                        onCompositionStart={[Function]}
                                        onCopy={[Function]}
                                        onCut={[Function]}
                                        onDragEnd={[Function]}
                                        onDragEnter={[Function]}
                                        onDragLeave={[Function]}
                                        onDragOver={[Function]}
                                        onDragStart={[Function]}
                                        onDrop={[Function]}
                                        onFocus={[Function]}
                                        onInput={[Function]}
                                        onKeyDown={[Function]}
                                        onKeyPress={[Function]}
                                        onKeyUp={[Function]}
                                        onMouseUp={[Function]}
                                        onPaste={[Function]}
                                        onSelect={[Function]}
                                        role="textbox"
                                        spellCheck={true}
                                        style={
                                          Object {
                                            "outline": "none",
                                            "whiteSpace": "pre-wrap",
                                            "wordWrap": "break-word",
                                          }
                                        }
                                        suppressContentEditableWarning={true}
                                        tabIndex={undefined}>
                                        <div
                                          data-contents="true">
                                          <div
                                            className="RichTextEditor__block___2Vs_D RichTextEditor__paragraph___3NTf9"
                                            data-block={true}
                                            data-editor="4ulug"
                                            data-offset-key="fauvt-0-0">
                                            <div
                                              className="public-DraftStyleDefault-block public-DraftStyleDefault-ltr"
                                              data-offset-key="fauvt-0-0">
                                              <span
                                                data-offset-key="fauvt-0-0"
                                                style={Object {}}>
                                                <br
                                                  data-text="true" />
                                              </span>
                                            </div>
                                          </div>
                                        </div>
                                      </div>
                                    </div>
                                  </div>
                                </div>
                              </div>
                            </div>
                            <div
                              className="clear" />
                          </div>
                        </fieldset>
                      </div>
                      <div
                        className="gte_editor_fields">
                        <label
                          className="gte_label"
                          htmlFor="date">
                          Date Time:
                        </label>
                        <div
                          className={undefined}>
                          <input
                            id="date"
                            name="date"
                            onChange={[Function]}
                            type="date"
                            value="" />
                        </div>
                        <div
                          className="clear" />
                      </div>
                      <div
                        className="gte_editor_fields">
                        <label
                          className="gte_label"
                          htmlFor="image">
                          Image:
                        </label>
                        <div
                          className={undefined}>
                          <input
                            id="image"
                            name="image"
                            type="file"
                            value={undefined} />
                        </div>
                        <div
                          className="clear" />
                      </div>
                    </div>
                  </div>
                </form>
              </div>
            </div>
            <div
              className="gte_footer">
              <div
                className="gte_form_err" />
              <div
                className="gte_form_buttons">
                <button
                  className="btn"
                  data-action="create"
                  id="gte_sent_btn"
                  onClick={[Function]} />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="gte_popup_background fade_in"
    onClick={[Function]} />
</div>
`;
